File: python_lstm/src/preprocess.py
import typing as t

import torch
from torch.utils.data import Dataset, DataLoader, random_split

SOS_TOKEN = "<sos>"
EOS_TOKEN = "<eos>"
PAD_TOKEN = "<pad>"


class TranslationDataset(Dataset):
    def __init__(self, data_path: str, max_len: int = 10) -> None:
        self.data_path = data_path
        self.max_len = max_len
        self.src_sentences: t.Sequence[t.Sequence[str]] = []
        self.trg_sentences: t.Sequence[t.Sequence[str]] = []
        self.src_vocab: t.Mapping[str, int] = {PAD_TOKEN: 0, SOS_TOKEN: 1, EOS_TOKEN: 2}
        self.trg_vocab: t.Mapping[str, int] = {PAD_TOKEN: 0, SOS_TOKEN: 1, EOS_TOKEN: 2}
        self.encoded_src: torch.Tensor = None
        self.encoded_trg: torch.Tensor = None

        self._read_data()
        self._build_vocab()
        self._encode_sentences()

    def _read_data(self) -> None:
        with open(self.data_path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                    
                parts = line.split("\t")
                if len(parts) != 2:
                    print(f"Skipping malformed line: {line}")
                    continue

                en, ru = parts
                self.src_sentences.append(en.strip().split())
                self.trg_sentences.append(ru.strip().split())

    def _build_vocab(self) -> None:
        src_index = len(self.src_vocab)
        for sentence in self.src_sentences:
            for token in sentence:
                if token not in self.src_vocab:
                    self.src_vocab[token] = src_index
                    src_index += 1

        trg_index = len(self.trg_vocab)
        for sentence in self.trg_sentences:
            for token in sentence:
                if token not in self.trg_vocab:
                    self.trg_vocab[token] = trg_index
                    trg_index += 1

    def _encode_sentences(self) -> None:
        self.encoded_src = []
        self.encoded_trg = []
        
        for src, trg in zip(self.src_sentences, self.trg_sentences):
            src_ids = [self.src_vocab[SOS_TOKEN]] + \
                      [self.src_vocab[t] for t in src if t in self.src_vocab] + \
                      [self.src_vocab[EOS_TOKEN]]
            trg_ids = [self.trg_vocab[SOS_TOKEN]] + \
                      [self.trg_vocab[t] for t in trg if t in self.trg_vocab] + \
                      [self.trg_vocab[EOS_TOKEN]]

            src_ids = src_ids[:self.max_len]
            trg_ids = trg_ids[:self.max_len]
            src_ids += [self.src_vocab[PAD_TOKEN]] * (self.max_len - len(src_ids))
            trg_ids += [self.trg_vocab[PAD_TOKEN]] * (self.max_len - len(trg_ids))

            self.encoded_src.append(src_ids)
            self.encoded_trg.append(trg_ids)

        self.encoded_src = torch.tensor(self.encoded_src, dtype=torch.long)
        self.encoded_trg = torch.tensor(self.encoded_trg, dtype=torch.long)

    def __len__(self) -> int:
        return len(self.encoded_src)

    def __getitem__(self, idx: int) -> tuple[torch.Tensor, torch.Tensor]:
        return self.encoded_src[idx], self.encoded_trg[idx]


def get_train_test_loaders(
    data_path: str, 
    batch_size: int = 2, 
    split_ratio: float = 0.8
) -> tuple[DataLoader, DataLoader, t.Mapping[str, int], t.Mapping[str, int]]:
    dataset = TranslationDataset(data_path)
    total_samples = len(dataset)
    train_size = int(split_ratio * total_samples)
    test_size = total_samples - train_size
    
    train_dataset, test_dataset = random_split(dataset, [train_size, test_size])
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)
    
    return train_loader, test_loader, dataset.src_vocab, dataset.trg_vocab

File: rust_lstm/src/preprocess.rs
use std::collections::HashMap;
use std::fs;
use tch::{Tensor, Kind};

pub const PAD_TOKEN: &str = "<pad>";
pub const SOS_TOKEN: &str = "<sos>";
pub const EOS_TOKEN: &str = "<eos>";


pub struct Vocab {
    pub token_to_id: HashMap<String, i64>,
    pub id_to_token: Vec<String>,
}


impl Vocab {
    pub fn new() -> Self {
        Vocab {
            token_to_id: HashMap::new(),
            id_to_token: Vec::new(),
        }
    }
    
    pub fn register_token(&mut self, token: &str) {
        if !self.token_to_id.contains_key(token) {
            let idx = self.id_to_token.len() as i64;
            self.token_to_id.insert(token.to_string(), idx);
            self.id_to_token.push(token.to_string());
        }
    }
}


pub struct TranslationData {
    pub src_tensors: Vec<Tensor>,
    pub trg_tensors: Vec<Tensor>,
    pub src_vocab: Vocab,
    pub trg_vocab: Vocab,
}


pub fn load_data(data_path: &str, max_len: usize) -> TranslationData {
    let content = fs::read_to_string(data_path).expect("Failed to read data file");
    let mut src_sentences: Vec<Vec<String>> = Vec::new();
    let mut trg_sentences: Vec<Vec<String>> = Vec::new();

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        
        let parts: Vec<&str> = line.split('\t').collect();
        if parts.len() != 2 {
            println!("Skipping malformed line: {}", line);
            continue;
        }
        
        let src: Vec<String> = parts[0]
            .split_whitespace()
            .map(|s| s.to_string())
            .collect();
        let trg: Vec<String> = parts[1]
            .split_whitespace()
            .map(|s| s.to_string())
            .collect();
            
        src_sentences.push(src);
        trg_sentences.push(trg);
    }

    let mut src_vocab = Vocab::new();
    let mut trg_vocab = Vocab::new();

    for token in &[PAD_TOKEN, SOS_TOKEN, EOS_TOKEN] {
        src_vocab.register_token(token);
        trg_vocab.register_token(token);
    }
    
    for sentence in &src_sentences {
        for token in sentence {
            src_vocab.register_token(token);
        }
    }
    
    for sentence in &trg_sentences {
        for token in sentence {
            trg_vocab.register_token(token);
        }
    }

    let mut src_tensors = Vec::new();
    let mut trg_tensors = Vec::new();

    for (src, trg) in src_sentences.iter().zip(trg_sentences.iter()) {
        let src_ids = encode_sentence(src, &src_vocab, max_len);
        let trg_ids = encode_sentence(trg, &trg_vocab, max_len);
        
        let src_tensor = Tensor::f_from_slice(&src_ids).unwrap().to_kind(Kind::Int64);
        let trg_tensor = Tensor::f_from_slice(&trg_ids).unwrap().to_kind(Kind::Int64);
        
        src_tensors.push(src_tensor);
        trg_tensors.push(trg_tensor);
    }

    TranslationData {
        src_tensors,
        trg_tensors,
        src_vocab,
        trg_vocab,
    }
}


fn encode_sentence(sentence: &Vec<String>, vocab: &Vocab, max_len: usize) -> Vec<i64> {
    let sos = *vocab.token_to_id.get(SOS_TOKEN).unwrap();
    let eos = *vocab.token_to_id.get(EOS_TOKEN).unwrap();
    let pad = *vocab.token_to_id.get(PAD_TOKEN).unwrap();
    
    let mut ids = Vec::new();
    ids.push(sos);
    
    for token in sentence {
        if let Some(&id) = vocab.token_to_id.get(token) {
            ids.push(id);
        }
    }
    
    ids.push(eos);
    
    if ids.len() > max_len {
        ids.truncate(max_len);
    } else {
        while ids.len() < max_len {
            ids.push(pad);
        }
    }
    
    ids
}
